simulate/simulate.go

```diff
@@ -121,166 +132,360 @@ func changeBase(originalBase dna.Base) dna.Base {
+			exonSeq = seq[exonStart:exonEnd]
+			for p := 0; p < len(exonSeq); p++ {
+				var codon CodonExt
+				codon.Seq = make([]BaseExt, 3)
+				if p%3 == 0 {
+					codon.Seq[0] = exonSeq[p]
+					codon.Seq[1] = exonSeq[p+1]
+					codon.Seq[2] = exonSeq[p+2]
+					allCodons = append(allCodons, codon)
+				}
+			}
+		} else if gene.ExonFrames[exon] != 0 { //Frame != 0
+			startFrame := gene.ExonFrames[exon]
+			exonSeq = seq[(gene.ExonStarts[exon] + startFrame):gene.ExonEnds[exon]]
+			for p := 0; p < len(exonSeq); p++ {
+				var codon CodonExt
+				codon.Seq = make([]BaseExt, 3)
+				if p%3 == 0 {
+					codon.Seq[0] = exonSeq[p]
+					codon.Seq[1] = exonSeq[p+1]
+					codon.Seq[2] = exonSeq[p+2]
+					allCodons = append(allCodons, codon)
 				}
 			}
 		}
 	}
+	return allCodons
+}
+
+//BasesToBaseExt converts a slice of dna.Base to a slice of BaseExt
+func BasesToBaseExt(seq []dna.Base) []BaseExt {
+	Sequence := make([]BaseExt, len(seq))
+	for i := 0; i < len(seq); i++ {
+		Sequence[i].Base = seq[i]
+		Sequence[i].SeqPos = i
+	}
+	return Sequence
+}
 
+//BaseExtToBases converts a slice of BaseExt to a slice of dna.Base
+func BaseExtToBases(seq []BaseExt) []dna.Base {
+	var newSequence []dna.Base
+
+	for i := 0; i < len(seq); i++ { //counter to put things back in order
+		for j := 0; j < len(seq); j++ { //counter to check all seq bases before incrementing to look for the next base in the newSeq
+			if seq[j].SeqPos == i {
+				newSequence = append(newSequence, seq[j].Base)
+				i += 1
+			}
+		}
+	}
+	if len(newSequence) != len(seq) {
+		log.Fatal("Cannot find order of bases")
+	}
 	return newSequence
 }
 
-//func getOverlapCDS(gtf map[string]*gtf.Gene, pos int) (cds *gtf.CDS, start bool) { //only works on pos strand genes
-//	for _, gene := range gtf {
-//		for _, transcript := range gene.Transcripts {
-//			for exonNum, exon := range transcript.Exons {
-//				if exon.Cds != nil {
-//					if exon.Cds.Start-1 <= pos && pos <= exon.Cds.End-1 {
-//						return exon.Cds, exonNum == 0
-//					}
-//				}
-//			}
-//		}
-//	}
-//	return nil, false
-//}
+//CodonExtToBaseExt converts a slice of CodonExt to a slice of BaseExt
+func CodonExtToBaseExt(allCodons []CodonExt) []BaseExt {
+	var base BaseExt
+	var seq []BaseExt
+	for c := 0; c < len(allCodons); c++ {
+		for s := 0; s < 3; s++ {
+			base = allCodons[c].Seq[s]
+			seq = append(seq, base)
+		}
+	}
+	return seq
+}
+
+//CodonExtToCodon converts a since CodonExt to a pointer to a single dna.Codon
+func CodonExtToCodon(cE CodonExt) *dna.Codon {
+	var codon dna.Codon
+	codon.Seq = make([]dna.Base, 3)
+
+	for c := 0; c < 3; c++ {
+		codon.Seq[c] = cE.Seq[c].Base
+	}
+	return &codon
+}
+
+//CodonExtsToCodons converts a slice of CodonExt to a slice of dna.Codon
+func CodonExtsToCodons(cE []CodonExt) []dna.Codon {
+	var codons []dna.Codon
+	var codon dna.Codon
+
+	for ext := 0; ext < len(cE); ext++ {
+		for c := 0; c < 3; c++ {
+			codon.Seq[c] = cE[ext].Seq[c].Base
```

Comment:
I would have thought you would need to allocate codon.Seq before running this line.  Does it work?  Maybe I am missing something.

Reply:
i actually wrote that, but didn't end up using it, so it's untested. You're right though, it probably doesn't work. I'll add a test for it because I think it's worth keeping here in case